lib/honeybadger/config_sync_service.rb

```diff
@@ -0,0 +1,82 @@
+require 'honeybadger/check_in'
+module Honeybadger
+  class ConfigSyncService
+    def initialize(config)
+      @config = config
+      @check_ins_cache = nil
+    end
+
+    def sync_check_ins
+      check_in_configs = @config.get(:check_ins) || []
+      
+      return [] if check_in_configs.empty?
+
+      check_ins = check_in_configs.map do |cfg|
+        check_in = CheckIn.from_config(cfg)
+
+        check_in.validate!
+        check_in
+      end
+
+      check_unique_names(check_ins)
+
+      created_or_updated = sync_existing_check_ins(check_ins)
+      removed = sync_removed_check_ins(check_ins)
+
+      return created_or_updated + removed
+    end
+
+    private
+
+    def check_unique_names(check_ins)
+      names = check_ins.map(&:name)
+      dupes = names.find_all {|n| names.count(n) > 1}.uniq
+      raise Honeybadger::InvalidCheckinConfig.new("Check Ins need to have unique names. #{dupes.join(", ")} used multiple times.") if dupes.length > 0
+    end
+
+    def get_checkin_by_name(project_id, name)
+      @check_ins_cache ||= @config.backend.get_check_ins(project_id)
+      @check_ins_cache.find {|c| c.name == name }
+    end
+
+    def sync_existing_check_ins(check_ins = [])
+      return [] if check_ins.empty?
+      result = []
+      check_ins.each do |check_in|
+        remote_check_in = if check_in.id
+          @config.backend.get_check_in(check_in.project_id, check_in.id)
+        else
+          get_checkin_by_name(check_in.project_id, check_in.name)
+        end
+        if remote_check_in
+          unless remote_check_in == check_in
+            result << @config.backend.update_check_in(check_in.project_id, remote_check_in.id, check_in)
+          end
+        else
+          result << @config.backend.create_check_in(check_in.project_id, check_in)
+        end
+      end
+      result
+    end
+
+    def sync_removed_check_ins(check_ins)
+      return [] if check_ins.nil? || check_ins.empty?
```

Comment:
@subzero10 @halfbyte I like this guard and think it's essential that we don't delete people's existing check-ins if they run `honeybadger check_ins sync` without having `check_ins` config. That said, do we also provide any kind of warning for people who want to try this feature and have a lot of existing check-ins? For example, if I have an existing Honeybadger project with check-ins, and I try to use this feature, my first assumption will not be "I need to put _all_ of my check-ins into my local config before running this command, otherwise they will be deleted."

Obviously we should highlight this behavior in the docs, but I'm wondering if we can do anything else to prevent accidents. Some people have hundreds of check-ins. 😬

Reply:
> Some people have hundreds of check-ins.

This feature is exactly for these people! Imagine that clicks they had to go through until they created those checkins.

Here are some ideas (on top of the warning in the documentation):
- We could create a `honeybadger check-ins pull` command that you run once and populates your config file with all your check-ins.
- We could implement an `inactive` state for check-ins and mark check-ins as inactive instead of deleting them. This would allow people to easily _rollback_ if they accidentally disable check-ins.

Reply:
> Another idea would be to prompt for confirmation in the CLI, but I wouldn't recommend it since we encourage people to use the command as part of their deployment pipelines.

One option is to prompt by default, but have a flag to skip the prompt for deployments.

Reply:
I also like the `honeybadger check-ins pull` idea — not as a solution for this problem, but it would be handy when migrating to this (we could consider adding it later).

Reply:
Sounds good! So, to sum up:
- Adopt the CLI to prompt for confirmation before going ahead.
- Create a CLI option to skip the prompt, to be used with CI environments.
- When writing the documentation page, make sure to add a callout saying that using the CLI will overwrite any existing check-ins not defined in the configuration file.


I suggest that we decide on the prompt, so that we can apply it on all libraries, how about this one:
```
Warning: Proceeding will replace all your current check-ins with the ones defined in the configuration file. Do you want to continue? (y/n)
```
What do you think?

Reply:
@subzero10 @joshuap I've added that prompt (took the liberty to add a line break) and added a `--yes` command line switch to override the confirmation. Not sure that's the best switch for this but it seems to be somewhat common for these kinda things, but let me know if you want to have it changed.

Reply:
apt uses `-y`/`--yes` when installing packages. I think that works.

@stympy @rabidpraxis @roelbondoc @subzero10 does that make sense to you?